### PR TITLE
Walker refactor

### PR DESCRIPTION
**Pull request by krl**

### CODE REVIEW

**Review by ZER0 — CHANGES_REQUESTED** [review]
LGTM, but please address the nits below before merging. Ping me or request another review for the approval.

File: src/compound.rs [hunk 1]
```diff
@@ -3,179 +3,126 @@
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 //
 // Copyright (c) DUSK NETWORK. All rights reserved.
-
```
[the remark this answers is not in the document]
Comment (krl) [reply src/compound.rs:6]: fixed

File: tests/multiple.rs [hunk 2]
```diff
@@ -0,0 +1,103 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+use core::borrow::Borrow;
```
[the remark this answers is not in the document]
Comment (krl) [reply tests/multiple.rs:7]: fixed

File: src/annotations/mod.rs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    fn from_leaf(leaf: &Leaf) -> Self;
+}
+
+/// Trait for defining how to combine Annotations
+pub trait Combine<C, A>: Annotation<C::Leaf>
+where
+    C: Compound<A>,
+{
+    /// Combines multiple annotations
+    fn combine<'a>(node: &'a C) -> Self;
```
[the remark this answers is not in the document]
Comment (krl) [reply src/annotations/mod.rs]: fixed

File: src/compound.rs [hunk 4]
```diff
@@ -3,179 +3,126 @@
...
-use core::borrow::Borrow;
 use core::marker::PhantomData;
 
-use canonical::{Canon, Store};
-
-use crate::annotation::{Annotated, Annotation, Cardinality};
-use crate::branch::{Branch, Step, Walk};
-use crate::branch_mut::{BranchMut, StepMut, WalkMut};
+use crate::annotations::{Annotated, Annotation, Combine, WrappedAnnotation};
```
[the remark this answers is not in the document]
Comment (krl) [reply src/compound.rs]: fixed

File: src/annotations/unit.rs [hunk 5]
```diff
@@ -0,0 +1,24 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+#![allow(clippy::unused_unit)]
```
[the remark this answers is not in the document]
Comment (krl) [reply src/annotations/unit.rs:7]: fixed

**Review by ZER0 — APPROVED** [review]
As discussed we'll squash the commits in release_0.7 before merge with the default branch.

**Pull request merged**